Userland/Libraries/LibDiff/Generator.cpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, Mustafa Quraish <mustafa@cs.toronto.edu>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Generator.h"
+
+namespace Diff {
+
+Vector<Hunk> from_text(StringView const& old_text, StringView const& new_text)
+{
+    auto old_lines = old_text.lines();
+    auto new_lines = new_text.lines();
+
+    enum Direction {
+        Up,       // Added a new line
+        Left,     // Removed a line
+        Diagonal, // Line remained the same a line
+    };
+
+    // A single cell in the DP-matrix. Cell (i, j) represents the longest common
+    // sub-sequence of lines between old_lines[0 : i] and new_lines[0 : j].
+    struct Cell {
+        size_t length;
+        Direction direction;
+    };
+
+    auto index = [&](size_t i, size_t j) { return i + (old_lines.size() + 1) * j; };
+
+    // DP-matrix.
+    auto dp = Vector<Cell>();
+    dp.resize((old_lines.size() + 1) * (new_lines.size() + 1));
+
+    // Initialize the first row and column
+    for (size_t i = 0; i <= old_lines.size(); ++i)
+        dp[index(i, 0)] = { 0, Direction::Left };
+
+    for (size_t j = 0; j <= new_lines.size(); ++j)
+        dp[index(0, j)] = { 0, Direction::Up };
+
+    // Fill in the rest of the DP table
+    for (size_t i = 1; i <= old_lines.size(); ++i) {
+        for (size_t j = 1; j <= new_lines.size(); ++j) {
+            if (old_lines[i - 1] == new_lines[j - 1]) {
+                dp[index(i, j)] = { dp[index(i - 1, j - 1)].length + 1, Direction::Diagonal };
+            } else {
+                auto up = dp[index(i, j - 1)].length;
+                auto left = dp[index(i - 1, j)].length;
+                if (up > left)
+                    dp[index(i, j)] = { up, Direction::Up };
+                else
+                    dp[index(i, j)] = { left, Direction::Left };
```

Comment:
If you are going to use the `index(..)` lambda to index the array, why not go one step forward and have it just return the reference to the element in the array?

Reply:
Nice idea!